src/ORM/Association/BelongsToMany.php

```diff
@@ -1095,11 +1095,23 @@ protected function _collectJointEntities($sourceEntity, $targetEntities)
     protected function _buildQuery($options)
     {
         $name = $this->_junctionAssociationName();
+        $assoc = $this->target()->association($name);
+        $queryBuilder = false;
+
+        if (!empty($options['queryBuilder'])) {
+            $queryBuilder = $options['queryBuilder'];
+            unset($options['queryBuilder']);
+        }
+
         $query = $this->_buildBaseQuery($options);
+        $query->addDefaultTypes($assoc->target());
+
+        if ($queryBuilder) {
+            $query = $queryBuilder($query);
```

Comment:
It was, at the end of the parent's eagerLoad() method. That was too late for limiting the fields, but also the right moment, for the query already had the types information for its associations.

I had to move it here, where it is called before the association is attached, in order to limit the fields; but then had to manually set the types for the associated table myself.